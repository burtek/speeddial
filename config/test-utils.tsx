import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';
import type { FC, PropsWithChildren, ReactElement } from 'react';
import { Provider as StoreProvider } from 'react-redux';

import { store } from '@@data/index';
import { ThemedApp } from 'src/theme/themed-app';

const AllTheProviders: FC<PropsWithChildren> = ({ children }) => (
    <StoreProvider store={store}>
        <ThemedApp>
            {children}
        </ThemedApp>
    </StoreProvider>
);

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
// eslint-disable-next-line import/export
export * from '@testing-library/react';

// override render method
// eslint-disable-next-line import/export
export { customRender as render };
