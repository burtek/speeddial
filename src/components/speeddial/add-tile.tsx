import AddIcon from '@mui/icons-material/Add';
import { Card, CardContent, Divider, IconButton } from '@mui/material';
import { FolderPlus, LinkPlus } from 'mdi-material-ui';
import type { FC } from 'react';
import { useState, useCallback } from 'react';

import { useAppDispatch } from '@@data/index';
import { actions as speeddialActions, ROOT_SPEEDDIAL_ID } from '@@data/speeddial/slice';

export const AddNewTile: FC<Props> = ({ parentId }) => {
    const dispatch = useAppDispatch();

    const onAddLink = useCallback(
        () => {
            dispatch(speeddialActions.createLink({ parentId }));
        },
        [dispatch, parentId]
    );
    const onAddGroup = useCallback(
        () => {
            dispatch(speeddialActions.createGroup());
        },
        [dispatch]
    );

    const [isMouseOver, setIsMouseOver] = useState(false);
    const onMouseOver = useCallback(() => {
        setIsMouseOver(true);
    }, [setIsMouseOver]);
    const onMouseOut = useCallback(() => {
        setIsMouseOver(false);
    }, [setIsMouseOver]);

    const defaultContent = <AddIcon fontSize="large" htmlColor="grey" />;
    const mouseOverContent = (
        <>
            {parentId === ROOT_SPEEDDIAL_ID && (
                <>
                    <IconButton onClick={onAddLink}>
                        <LinkPlus fontSize="large" htmlColor="grey" />
                    </IconButton>
                    <Divider flexItem variant="middle" />
                </>
            )}
            <IconButton onClick={onAddGroup}>
                <FolderPlus fontSize="large" htmlColor="grey" />
            </IconButton>
        </>
    );

    return (
        <Card
            variant="outlined"
            onMouseOver={onMouseOver}
            onMouseOut={onMouseOut}
            sx={{ cursor: 'pointer', width: 150, display: 'flex', justifyContent: 'center', alignItems: 'stretch' }}
        >
            <CardContent
                sx={{
                    'display': 'flex',
                    'justifyContent': 'center',
                    'alignItems': 'center',
                    'flexDirection': 'column',
                    'p': 0,
                    '&:last-child': { p: 0 },
                    'flex': 1
                }}
            >
                {isMouseOver ? mouseOverContent : defaultContent}
            </CardContent>
        </Card>
    );
};

interface Props {
    parentId: string;
}
