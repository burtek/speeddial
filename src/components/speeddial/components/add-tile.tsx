import AddIcon from '@mui/icons-material/Add';
import { Card, CardContent, Divider, IconButton, styled } from '@mui/material';
import { FolderPlus, LinkPlus } from 'mdi-material-ui';
import type { FC } from 'react';
import { useState, useCallback } from 'react';
import { useTranslation } from 'react-i18next';

import { useAppDispatch } from '@@data/redux-toolkit';
import { actions as speeddialActions } from '@@data/speeddial/slice';

import { TILE_WIDTH } from './_constants';


const AddTileWrapper = styled(Card)({
    cursor: 'pointer',
    width: TILE_WIDTH,
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'stretch'
});

const AddTileContent = styled(CardContent)({
    'display': 'flex',
    'justifyContent': 'center',
    'alignItems': 'center',
    'flexDirection': 'column',
    'padding': 0,
    '&:last-child': { padding: 0 },
    'flex': 1
});

export const AddNewTile: FC<Props> = ({ parentId, hideAddGroup }) => {
    const { t } = useTranslation();

    const dispatch = useAppDispatch();

    const onAddLink = useCallback(() => {
        dispatch(speeddialActions.createLink({ parentId }));
    }, [dispatch, parentId]);
    const onAddGroup = useCallback(() => {
        dispatch(speeddialActions.createGroup());
    }, [dispatch]);

    const [isMouseOver, setIsMouseOver] = useState(false);
    const onMouseOver = useCallback(() => {
        setIsMouseOver(true);
    }, [setIsMouseOver]);
    const onMouseOut = useCallback(() => {
        setIsMouseOver(false);
    }, [setIsMouseOver]);

    const defaultContent = (
        <AddIcon
            aria-label={hideAddGroup ? t('tooltips.add_link') : t('tooltips.add_link_or_group')}
            fontSize="large"
            htmlColor="grey"
        />
    );
    const mouseOverContent = (
        <>
            <IconButton
                aria-label={t('tooltips.add_link')}
                onClick={onAddLink}
                title={t('tooltips.add_link')}
            >
                <LinkPlus fontSize="large" htmlColor="grey" />
            </IconButton>
            {hideAddGroup
                ? null
                : (
                    <>
                        <Divider flexItem variant="middle" />
                        <IconButton
                            aria-label={t('tooltips.add_group')}
                            onClick={onAddGroup}
                            title={t('tooltips.add_group')}
                        >
                            <FolderPlus fontSize="large" htmlColor="grey" />
                        </IconButton>
                    </>
                )}
        </>
    );

    return (
        <AddTileWrapper
            variant="outlined"
            onMouseOver={onMouseOver}
            onMouseOut={onMouseOut}
        >
            <AddTileContent>
                {isMouseOver ? mouseOverContent : defaultContent}
            </AddTileContent>
        </AddTileWrapper>
    );
};

interface Props {
    parentId: string;
    hideAddGroup?: boolean;
}
