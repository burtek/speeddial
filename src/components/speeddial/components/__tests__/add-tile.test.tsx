import userEvent from '@testing-library/user-event';

import { render, screen } from '@@config/test-utils';

import { AddNewTile } from '../add-tile';


describe('speeddial/add-tile', () => {
    beforeEach(() => {
        vitest.clearAllMocks();
    });

    it.each([
        {
            expectedNotToExist: ['tooltips.add_link_or_group', 'tooltips.add_group'],
            hover: 'tooltips.add_link',
            onAddLink: vitest.fn(),
            onAddGroup: undefined
        },
        {
            expectedNotToExist: ['tooltips.add_link', 'tooltips.add_group'],
            hover: 'tooltips.add_link_or_group',
            onAddLink: vitest.fn(),
            onAddGroup: vitest.fn()
        }
    ])('should render add icon without button', ({ expectedNotToExist, hover, onAddGroup, onAddLink }) => {
        const { container } = render(<AddNewTile onAddLink={onAddLink} onAddGroup={onAddGroup} />);

        expect(screen.queryAllByRole('button')).toHaveLength(0);
        expect(container.querySelectorAll('svg')).toHaveLength(1);
        expect(screen.getByLabelText(hover)).toBeInTheDocument();

        expectedNotToExist.forEach(label => {
            expect(screen.queryByLabelText(label)).not.toBeInTheDocument();
        });

        expect(container).toMatchSnapshot();
    });

    it.each([
        {
            hover: 'tooltips.add_link_or_group',
            onAddLink: vitest.fn(),
            onAddGroup: vitest.fn(),
            expectedButtonLabels: ['tooltips.add_link', 'tooltips.add_group'],
            notExpectedButtonLabels: ['tooltips.add_link_or_group']
        },
        {
            hover: 'tooltips.add_link',
            onAddLink: vitest.fn(),
            onAddGroup: undefined,
            expectedButtonLabels: ['tooltips.add_link'],
            notExpectedButtonLabels: ['tooltips.add_link_or_group', 'tooltips.add_group']
        }
    ])('should render expected buttons on hover', async ({ hover, expectedButtonLabels, notExpectedButtonLabels, ...props }) => {
        const user = userEvent.setup();
        const { container } = render(<AddNewTile {...props} />);

        await user.hover(screen.getByLabelText(hover));
        const buttons = screen.getAllByRole('button');

        expect(buttons).toHaveLength(expectedButtonLabels.length);

        expectedButtonLabels.forEach(label => {
            expect(screen.getByLabelText(label)).toBeInTheDocument();
        });
        notExpectedButtonLabels.forEach(label => {
            expect(screen.queryByLabelText(label)).not.toBeInTheDocument();
        });

        expect(container).toMatchSnapshot();
    });

    describe('callbacks', () => {
        const onAddLink = vitest.fn();
        const onAddGroup = vitest.fn();

        it.each([
            {
                clickLabel: 'tooltips.add_link',
                expectation: () => {
                    expect(onAddLink).toHaveBeenCalled();
                }
            },
            {
                clickLabel: 'tooltips.add_group',
                expectation: () => {
                    expect(onAddGroup).toHaveBeenCalled();
                }
            }
        ])('$clickLabel should call back expected function', async ({ clickLabel, expectation }) => {
            const user = userEvent.setup();
            render(<AddNewTile onAddLink={onAddLink} onAddGroup={onAddGroup} />);

            await user.hover(screen.getByLabelText('tooltips.add_link_or_group'));
            const button = await screen.findByLabelText(clickLabel);

            expect(button).toBeInTheDocument();

            await user.click(button);

            expectation();
        });
    });
});
