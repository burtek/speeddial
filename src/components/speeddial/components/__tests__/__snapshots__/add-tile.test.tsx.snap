// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`speeddial/add-tile > should render add icon without button 1`] = `
.emotion-0 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.12);
  overflow: hidden;
  cursor: pointer;
  width: 150px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.emotion-1 {
  padding: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-1:last-child {
  padding-bottom: 24px;
}

.emotion-1:last-child {
  padding: 0;
}

.emotion-2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 2.1875rem;
}

<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root emotion-0"
  >
    <div
      class="MuiCardContent-root emotion-1"
    >
      <svg
        aria-hidden="true"
        aria-label="tooltips.add_link"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge emotion-2"
        color="grey"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`speeddial/add-tile > should render add icon without button 2`] = `
.emotion-0 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.12);
  overflow: hidden;
  cursor: pointer;
  width: 150px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.emotion-1 {
  padding: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-1:last-child {
  padding-bottom: 24px;
}

.emotion-1:last-child {
  padding: 0;
}

.emotion-2 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 2.1875rem;
}

<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root emotion-0"
  >
    <div
      class="MuiCardContent-root emotion-1"
    >
      <svg
        aria-hidden="true"
        aria-label="tooltips.add_link_or_group"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge emotion-2"
        color="grey"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`speeddial/add-tile > should render expected buttons on hover 1`] = `
.emotion-0 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.12);
  overflow: hidden;
  cursor: pointer;
  width: 150px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.emotion-1 {
  padding: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-1:last-child {
  padding-bottom: 24px;
}

.emotion-1:last-child {
  padding: 0;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2::-moz-focus-inner {
  border-style: none;
}

.emotion-2.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-2 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-2:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-2:hover {
    background-color: transparent;
  }
}

.emotion-2.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 2.1875rem;
}

.emotion-4 {
  margin: 0;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-width: 0;
  border-style: solid;
  border-color: rgba(0, 0, 0, 0.12);
  border-bottom-width: thin;
  margin-left: 16px;
  margin-right: 16px;
  -webkit-align-self: stretch;
  -ms-flex-item-align: stretch;
  align-self: stretch;
  height: auto;
}

<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root emotion-0"
  >
    <div
      class="MuiCardContent-root emotion-1"
    >
      <button
        aria-label="tooltips.add_link"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-2"
        tabindex="0"
        title="tooltips.add_link"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge emotion-3"
          color="grey"
          data-testid="LinkPlusIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,7H11V9H7A3,3 0 0,0 4,12A3,3 0 0,0 7,15H11V17H7A5,5 0 0,1 2,12A5,5 0 0,1 7,7M17,7A5,5 0 0,1 22,12H20A3,3 0 0,0 17,9H13V7H17M8,11H16V13H8V11M17,12H19V15H22V17H19V20H17V17H14V15H17V12Z"
          />
        </svg>
      </button>
      <hr
        class="MuiDivider-root MuiDivider-middle MuiDivider-flexItem emotion-4"
      />
      <button
        aria-label="tooltips.add_group"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-2"
        tabindex="0"
        title="tooltips.add_group"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge emotion-3"
          color="grey"
          data-testid="FolderPlusIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M13 19C13 19.34 13.04 19.67 13.09 20H4C2.9 20 2 19.11 2 18V6C2 4.89 2.89 4 4 4H10L12 6H20C21.1 6 22 6.89 22 8V13.81C21.12 13.3 20.1 13 19 13C15.69 13 13 15.69 13 19M20 18V15H18V18H15V20H18V23H20V20H23V18H20Z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`speeddial/add-tile > should render expected buttons on hover 2`] = `
.emotion-0 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  border: 1px solid rgba(0, 0, 0, 0.12);
  overflow: hidden;
  cursor: pointer;
  width: 150px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.emotion-1 {
  padding: 16px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 0;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-1:last-child {
  padding-bottom: 24px;
}

.emotion-1:last-child {
  padding: 0;
}

.emotion-2 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2::-moz-focus-inner {
  border-style: none;
}

.emotion-2.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-2 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-2:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-2:hover {
    background-color: transparent;
  }
}

.emotion-2.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 2.1875rem;
}

<div>
  <div
    class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root emotion-0"
  >
    <div
      class="MuiCardContent-root emotion-1"
    >
      <button
        aria-label="tooltips.add_link"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium emotion-2"
        tabindex="0"
        title="tooltips.add_link"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge emotion-3"
          color="grey"
          data-testid="LinkPlusIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7,7H11V9H7A3,3 0 0,0 4,12A3,3 0 0,0 7,15H11V17H7A5,5 0 0,1 2,12A5,5 0 0,1 7,7M17,7A5,5 0 0,1 22,12H20A3,3 0 0,0 17,9H13V7H17M8,11H16V13H8V11M17,12H19V15H22V17H19V20H17V17H14V15H17V12Z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
